### DISCUSSION TITLE
How to get the timestamp of a sample?

### DISCUSSION THREAD

**Discussion by chisarie** [eclipse-zenoh/roadmap#108] [Q&A]
Hi, I am looking at Zenoh for the first time today and it looks really cool! One question I have is how can I get the timestamp of a certain message. I am following the simple examples from the documentation, in python:

```python
# Publisher
pub = session.declare_publisher(key)
value = np.random.randint(low=15, high=30, size=(2, 2))
pub.put(value.tobytes())
```

```python
def listener_callback(sample: zenoh.Sample):
    # To decode back the numpy array, we need to know its shape and dtype
    value = np.frombuffer(sample.payload, dtype=np.int64).reshape(2, 2)
    print(f"Received {sample.kind} ('{sample.key_expr}': '{value}')")
    print(f"Timestamp: {sample.timestamp}")  # Unfortunately this is None

# Subscriber
sub = session.declare_subscriber(key, listener_callback)
```

Unfortunately, `sample.timestamp` is `None`. How can I get this field? 
Thank you very much in advance!

**Comment by gabrik** [comment]
Hi @chisarie,

In the default Zenoh configuration the timestamps are added by routers, I guess in your case you have a peer-to-peer topology without any router.
If that's the case you need to enable timestamping in `client/peers`, by using the following configuration.

```
{
timestamping: {
    /// Whether data messages should be timestamped if not already.
    /// Accepts a single boolean value or different values for router, peer and client.
    enabled: { router: true, peer: true, client: true },
}
```
Source: https://github.com/eclipse-zenoh/zenoh/blob/master/DEFAULT_CONFIG.json5#L75

With thin configuration `sample.timestamp` should return you the information you are looking for.

**Comment by chisarie** [comment]
Thank you very much Gabriele, it works!

I have one additional question: what is the recommended way to handle a large numpy array whose shape is not known a priori?
Is it possible to pass the shape of the array as additional metadata when publishing, so that the subscriber is able to use the information to decode the array? Something like the following pseudocode:

```python
# Publisher
pub = session.declare_publisher(key)
value = np.random.randint(low=15, high=30, size=(2, 2))
pub.put(value.tobytes(), metadata={"shape": value.shape})

# Subscriber
def callback(sample: zenoh.Sample):
    # To decode back the numpy array, we need to know its shape and dtype
    value = np.frombuffer(sample.payload, dtype=np.int64).reshape(sample.metadata.shape)
    print(f"Received {sample.kind} ('{sample.key_expr}': '{value}')")

sub = session.declare_subscriber(key, callback)
```

Thank you again :)
